### ISSUE TITLE
[Test]Simplify the testbed for basic precheckin test

### ISSUE THREAD

**Issue by shaominchen** [vmware/docker-volume-vsphere#1515]
Currently the precheckin tests (i.e. "test-all" target) requires a setup with at least 3 nodes deployed in swarm cluster mode. This is too heavy for a basic precheckin test. For internal developers, we can use nimbus to deploy a testbed with one single command. But for external developers, it will be too complicated to set up the environment simply for running the precheckin test.

On the other hand, we also want to make sure our CI can capture most common issues by running all important e2e test cases. Also we should still provide option to run any complicated test cases locally with some optional parameters.

To achieve these different goals, here's our plan:

1. Developers should be able to run test-all on a single VM. This is achieved by moving all extensive test cases (which require complicated setup) to one or more special labels
2. Developers should be able to run complicated test case by specifying some optional parameters. The assumption is that the environment should be ready for running those tests
3. CI should continue to run all extensive test cases as we have already planned

@msterin @shuklanirdesh82

**Comment by shuklanirdesh82** [comment]
@ashahi1, @shaominchen & I discussed following approach while invoking `make test-all` locally.

1. Basic testbed (1 ESX + 1 VM) : run tests eligible for such configuration and print out nice message about test run and suggesting to use recommended testbed.

2. Recommended testbed (1 ESX + 2 VMs): makes sure required parameters (env. variables) are set correctly and run tests otherwise stop the execution immediately before running any test.

3. Special need like Swarm testbed (as of now): Such specific test should assert out if the test bed need doesn't satisfy without blocking other test run. Such test setup should provide sufficient error message while asserting out (should not use `log.fatal()`)

**Comment by govint** [comment]
@shuklanirdesh82 - is this issue about identifying tests that are good to run with 1 VM and 1 ESX? Make a separate target for those to be run?

There are two issues here::

1. Using 1VM + 1ESX vs. 1ESX + 3VMs - I'm not sure why this is needed, a developer who can create 1VM + 1ESX can create two more VMs, set up a swarm and run their tests. And most likely those set ups will live for a long enough to be reused.

2. Identifying a "basic test set" - The issue with identifying a set of tests as "basic" is that soon they aren't enough to validate the code being checked in. Dev. is submitting code that the pre-checkin or basic tests aren't testing.

We can instead identify a set for regression that must be run before posting to CI perhaps. That set can exercise core workflows.

Even then, reducing the testbed to 1VM + 1ESX seems too constraining and I'm not sure of the value.

**Comment by shuklanirdesh82** [comment]
> Even then, reducing the testbed to 1VM + 1ESX seems too constraining and I'm not sure of the value.

> If anything we can create a script that will set up the testbed for the user and then allow the user to customize for their env.

Thanks @govint for bringing this point!

@msterin @shaominchen and I discussed about following approach that we should do without compromising local testing effort. As you pointed out sticking with 1ESX +3VM there are 2 different areas to achieve the plan:

- Area 1: Testbed automation for an external contributor
- Area 2: Test cases categorization

<br/> 

- **Area 1:** Testbed automation for an external contributor

**Testbed automation**: Having an automation testbed deployment approach for an external contributor (no issue for inside contributor as we are using nimbus infrastructure to deploy our testbed)

1. **When there is only ESX already deployed/available**: Automation script should ask about the ESX IP to create 3 VMs from the preconfigured OVF/OVA. We may create preconfigured photon/ubuntu OVA and placed public storage location so our script can configure the testbed to satisfy `make test-all` requirement (configures swarm cluster automatically as well)

2. Having a **strict requirement** for a user to deploy 1 ESX + 3 VMs: We can have a script that creates swarm cluster automatically to invoke `make test-all`

Note: for this approach, script configures swarm cluster only; for any other configuration requirement our CONTRIBUTING.md should call out explicitly.

- **Area 2:** Test cases categorization

Ideally we should invoke `make test-all` before generating PR but there should be some support to exclude test/test_group that user don't want to run in case of not having ideal testbed at first point or for any other reason it does not want to run. 

e.g. 

1. I would like to exclude swarm tests or failover/destructive test.
`make test-all --exclude swarm_tests` or `make test-all --exclude destructive_tests` 

1. I would like to run only basic test
`make test-all --include swarm_tests`

In any case CI is running all tests, user will be having local testbed to investigate further in case of PR failure on CI.

**REQUIRED**: In any case our CONTRIBUTING.md should be having all detailed instruction about the approach we are proposing.

/CC @shaominchen @msterin

**Comment by govint** [comment]
We could work with the photon team to help them create the OVA with the plugin installed, but I don't think we should be maintaining OVAs.

I'll work on creating the test bed with inputs as:
1. ESX IP available - create the swarm cluster of 1ESX and 3 VMs, with plugin installed on each and the service on the ESX host.
1. Cluster already available - verify that docker swarm cluster is present and install the plugin

**Comment by govint** [comment]
I'm done with the setup part of this issue and will post the PR tomorrow.  This allows the user to run a make target to set up a config of 1 master + 2 workers nodes with docker swarm initialized (can be excluded if needed, leaving just three nodes with docker installed on those) and creates a manifest of the master/worker IPs and env. settings that the user can init into their env and run the tests.

The approach is to use govc to import the OVAs (which the set up script can download from a URL or use an OVA path on the local host) and then create the cluster on the VMs. So, user as usual sets GOVC_* variables in their env. before running make for the setup target. User also specifies which datastore to use to import the OVAs into on the ESX host.

Investigated using ovftool, but thats licensed and I don't think we want to include it in the docker image used to run these make ops. Using govc as its already in place and gets us the same function.

Identifying the tests to run will be a separate PR.

**Comment by govint** [comment]
Depends on #1612

**Comment by govint** [comment]
Moving this to future, deploying a couple of VMs and an esx and setting up swarm is easily doable and there are multiple tools that already allow a user to do that. I've got an implementation that uses govc to talk to the ESX host to create a setup, but its more involved than using VC directly or the vsphere client for that matter to deploy OVAs to an ESX host. I'm not seeing much benefit adding code (scripts) to the plugin code base to create VM infra to run e2e tests. Moving to future unless this is needed earlier.

**Comment by govint** [comment]
Testbed set up is best done as per tools that the developer has and is familiar with. Creating a test setup seems unnecessary in the vDVS code. Its a good idea but its already addressed by the many tools already available to deploy and config a set up on vSphere.
